src/macros.jl

```diff
@@ -1005,34 +1005,178 @@ function parsevariable(_error::Function, infoexpr::VariableInfoExpr, lvalue, lsi
+* `start`: Sets the variable starting value used as initial guess in optimization.
+* `binary`: Sets whether the variable is binary or not.
+* `integer`: Sets whether the variable is integer or not.
+* `variabletype`: See the "Note for extending the variable macro " section below.
+
+## Variable naming
+
+There a two different aspects of the variable naming that needs to be
+distinguished when created a scalar variable (resp. a container of variables)
+
+* The name of the local variable created (if any) holding the variable (resp.
+  the container of variables) which corresponds to the name that can be used
+  to retrieve it using `m[:varname]`.
+* The name of the variable (resp. each variable in the container) used for
+  printing. This corresponds to the [`MathOptInterface.VariableName`](@ref)
+  attribute.
+
+When creating a variable using the syntax `@variable(m; kwargs...)` or when
+`varexpr` is of the form `[...]`, we say that the variable is *anonymous*.
+For anonymous variables, no local variable is created holding the result
+and it is not stored in the model, i.e. it is not possible to retrieve the
+variable using `m[:varname]`.
+
+Otherwise, when the variable is not anonymous, the name used both for the
+local variable created and the key for retrieving the variable in the model
+are determined from `varexpr`. The name is `varname` when `varexpr` is
+`varname` or `varname[...]`.
+
+The name of the variable is based on the base name which is specified by the
+`basename` keyword argument. When the `basename` keyword argument is not
+specified, the name depends on whether the variable is anonymous:
+
+* if the variable is anonymous, then the
+  [`MathOptInterface.VariableName`](@ref) attribute is not set and the name
+  used for printing is `noname`.
+* otherwise, the base name is set to `varname`.
+
+The name of the variables set to the [`MathOptInterface.VariableName`](@ref)
+attribute and used for printing is then `basename` for scalar variables
+and `basename[i1,i2,...,in]` for the variable at indices `i1`, `i2`, ..., `in`
+in a container of variables.
+
+## Note for extending the variable macro
+
+The single scalar variable or each scalar variable of the container are created
+using `addvariable(m, buildvariable(_error, info, extra_args...;
+extra_kwargs...))` where
+
+* `m` is the model passed to the `@variable` macro;
+* `_error` is an error function showing the `@variable` call in addition to the
+  error message given as argument;
+* `info` is the `VariableInfo` struct containing the information gathered in
+  `expr`, the recognized keyword arguments (except `basename` and
+  `variabletype`) and the recognized positional arguments (except `Symmetric`
+  and `PSD`);
+* `extra_args` are the unrecognized positional arguments of `args` plus the
+  value of the `variabletype` keyword argument if present. The `variabletype`
+  keyword argument allows to pass a positional argument to `buildvariable`
```

Comment:
Nit: allows who? Perhaps: "keyword argument allows the user to pass a position argument to?